plugins/ecs-bridge/engine/engine.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package engine
+
+import (
+	"net"
+	"syscall"
+
+	"github.com/aws/amazon-ecs-cni-plugins/pkg/cniipamwrapper"
+	"github.com/aws/amazon-ecs-cni-plugins/pkg/cniipwrapper"
+	"github.com/aws/amazon-ecs-cni-plugins/pkg/cninswrapper"
+	"github.com/aws/amazon-ecs-cni-plugins/pkg/netlinkwrapper"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+)
+
+// zeroLengthIPString is what we expect net.IP.String() to return if the
+// ip has length 0. We use this to determing if an IP is empty.
+// Refer https://golang.org/pkg/net/#IP.String
+const zeroLengthIPString = "<nil>"
+
+// Engine represents the execution engine for the ECS Bridge plugin.
+// It defines all the operations performed during the execution of the
+// plugin
+type Engine interface {
+	CreateBridge(bridgeName string, mtu int) (*netlink.Bridge, error)
+	CreateVethPair(netnsName string, mtu int, interfaceName string) (*current.Interface, string, error)
+	AttachHostVethInterfaceToBridge(hostVethName string, bridge *netlink.Bridge) (*current.Interface, error)
+	RunIPAMPluginAdd(plugin string, netConf []byte) (*current.Result, error)
+	ConfigureContainerVethInterface(netnsName string, result *current.Result, interfaceName string) error
+	ConfigureBridge(result *current.Result, bridge *netlink.Bridge) error
+	GetInterfaceIPV4Address(netnsName string, interfaceName string) (string, error)
+	RunIPAMPluginDel(plugin string, netconf []byte) error
+	DeleteVeth(netnsName string, interfaceName string) error
+}
+
+type engine struct {
+	netLink netlinkwrapper.NetLink
+	ns      cninswrapper.NS
+	ip      cniipwrapper.IP
+	ipam    cniipamwrapper.IPAM
+}
+
+// New creates a new Engine object
+func New() Engine {
+	return &engine{
+		netLink: netlinkwrapper.NewNetLink(),
+		ns:      cninswrapper.NewNS(),
+		ip:      cniipwrapper.New(),
+		ipam:    cniipamwrapper.New(),
+	}
+}
+
+// CreateBridge creates the bridge if needed
+func (engine *engine) CreateBridge(bridgeName string, mtu int) (*netlink.Bridge, error) {
+	bridgeLinkAttributes := netlink.NewLinkAttrs()
+	bridgeLinkAttributes.MTU = mtu
+	bridgeLinkAttributes.Name = bridgeName
+
+	bridge := &netlink.Bridge{
+		LinkAttrs: bridgeLinkAttributes,
+	}
+
+	err := engine.netLink.LinkAdd(bridge)
+	if err != nil && err != syscall.EEXIST {
```

Comment:
Thinking more on this and after playing a bit with the code, I think we need support from `netlink.LinkByName` to do this correctly. We need it to return a named `LinkNotFound"` error instead of `fmt.Errorf()` stuff that it does today. I'll come back to this once we fix that upstream.